File: Simplified/NYPLSettings.m
#import "NYPLSettings.h"
#import "NSDate+NYPLDateAdditions.h"
#import "NYPLBook.h"
#import "NYPLBookRegistry.h"
#import "NYPLConfiguration.h"
#import "SimplyE-Swift.h"

const NSInteger Version = 1;

NSString *const NYPLSettingsDidChangeNotification = @"NYPLSettingsDidChangeNotification";
NSString *const NYPLCurrentAccountDidChangeNotification = @"NYPLCurrentAccountDidChangeNotification";
NSString *const NYPLSyncBeganNotification = @"NYPLSyncBeganNotification";
NSString *const NYPLSyncEndedNotification = @"NYPLSyncEndedNotification";

static NSString *const customMainFeedURLKey = @"NYPLSettingsCustomMainFeedURL";

static NSString *const accountMainFeedURLKey = @"NYPLSettingsAccountMainFeedURL";

static NSString *const renderingEngineKey = @"NYPLSettingsRenderingEngine";

static NSString *const legacyUserAcceptedEULAKey = @"NYPLSettingsUserAcceptedEULA";

//static NSString *const settingsSynchronizeAnnotationsKey = @"NYPLSettingsSynchronizeAnnotationsKey";

static NSString *const userPresentedAgeCheckKey = @"NYPLUserPresentedAgeCheckKey";

static NSString *const userSeenFirstTimeSyncMessageKey = @"userSeenFirstTimeSyncMessageKey";

static NSString *const currentCardApplicationSerializationKey = @"NYPLSettingsCurrentCardApplicationSerialized";

static NSString *const settingsLibraryAccountsKey = @"NYPLSettingsLibraryAccountsKey";

static NSString *const settingsOfflineQueueKey = @"NYPLSettingsOfflineQueueKey";

static NSString *const settingsAnnotationsOfflineQueueKey = @"NYPLSettingsAnnotationsOfflineQueueKey";

static NSString *const versionKey = @"NYPLSettingsVersionKey";


static NYPLSettingsRenderingEngine RenderingEngineFromString(NSString *const string)
{
  if(!string || [string isEqualToString:@"automatic"]) {
    return NYPLSettingsRenderingEngineAutomatic;
  }
  
  if([string isEqualToString:@"readium"]) {
    return NYPLSettingsRenderingEngineReadium;
  }
  
  @throw NSInvalidArgumentException;
}

static NSString *StringFromRenderingEngine(NYPLSettingsRenderingEngine const renderingEngine)
{
  switch(renderingEngine) {
    case NYPLSettingsRenderingEngineAutomatic:
      return @"automatic";
    case NYPLSettingsRenderingEngineReadium:
      return @"readium";
  }
}

@implementation NYPLSettings

+ (NYPLSettings *)sharedSettings
{
  static dispatch_once_t predicate;
  static NYPLSettings *sharedSettings;
  
  dispatch_once(&predicate, ^{
    sharedSettings = [[self alloc] init];
    if(!sharedSettings) {
      NYPLLOG(@"Failed to create shared settings.");
    }
  });
  
  return sharedSettings;
}

- (NSURL *)customMainFeedURL
{
  return [[NSUserDefaults standardUserDefaults] URLForKey:customMainFeedURLKey];
}

- (NSURL *)accountMainFeedURL
{
  return [[NSUserDefaults standardUserDefaults] URLForKey:accountMainFeedURLKey];
}

- (BOOL) userHasSeenWelcomeScreen
{
  return [[NSUserDefaults standardUserDefaults] boolForKey:userHasSeenWelcomeScreenKey];
}

- (BOOL) userPresentedAgeCheck
{
  return [[NSUserDefaults standardUserDefaults] boolForKey:userPresentedAgeCheckKey];
}

- (BOOL) userHasSeenFirstTimeSyncMessage
{
  return [[NSUserDefaults standardUserDefaults] boolForKey:userSeenFirstTimeSyncMessageKey];
}

// FIXME: This should be in `AccountsManager`, not `NYPLSettings`.
- (NSArray *) settingsAccountsList
{
  NSArray *libraryAccounts = [[NSUserDefaults standardUserDefaults] arrayForKey:settingsLibraryAccountsKey];
  // If user has not selected any accounts yet, return the "currentAccount"
  if (!libraryAccounts) {
    NSString *currentLibrary = [AccountsManager shared].currentAccount.uuid;
    [self setSettingsAccountsList:@[currentLibrary, [AccountsManager NYPLAccountUUIDs][2]]];
    return [self settingsAccountsList];
  } else {
    return libraryAccounts;
  }
}

- (NYPLCardApplicationModel *)currentCardApplication
{
  NSData *currentCardApplicationSerialization = [[NSUserDefaults standardUserDefaults] objectForKey:currentCardApplicationSerializationKey];
  if (!currentCardApplicationSerialization)
    return nil;
  
  return [NSKeyedUnarchiver unarchiveObjectWithData:currentCardApplicationSerialization];
}

- (void)setUserHasSeenWelcomeScreen:(BOOL)userPresentedScreen
{
  [[NSUserDefaults standardUserDefaults] setBool:userPresentedScreen forKey:userHasSeenWelcomeScreenKey];
  [[NSUserDefaults standardUserDefaults] synchronize];
}

- (void)setUserPresentedAgeCheck:(BOOL)userPresentedAgeCheck
{
  [[NSUserDefaults standardUserDefaults] setBool:userPresentedAgeCheck forKey:userPresentedAgeCheckKey];
  [[NSUserDefaults standardUserDefaults] synchronize];
}

- (void)setUserHasSeenFirstTimeSyncMessage:(BOOL)seenSyncMesssage
{
  [[NSUserDefaults standardUserDefaults] setBool:seenSyncMesssage forKey:userSeenFirstTimeSyncMessageKey];
  [[NSUserDefaults standardUserDefaults] synchronize];
}

- (void)setCustomMainFeedURL:(NSURL *const)customMainFeedURL
{
  if(!customMainFeedURL && !self.customMainFeedURL) return;
  if([customMainFeedURL isEqual:self.customMainFeedURL]) return;
  
  [[NSUserDefaults standardUserDefaults] setURL:customMainFeedURL forKey:customMainFeedURLKey];
  [[NSUserDefaults standardUserDefaults] synchronize];
  
  [[NSNotificationCenter defaultCenter]
   postNotificationName:NYPLSettingsDidChangeNotification
   object:self];
}
- (void)setAccountMainFeedURL:(NSURL *const)accountMainFeedURL
{
  if(!accountMainFeedURL && !self.accountMainFeedURL) return;
  if([accountMainFeedURL isEqual:self.accountMainFeedURL]) return;
  
  [[NSUserDefaults standardUserDefaults] setURL:accountMainFeedURL forKey:accountMainFeedURLKey];
  [[NSUserDefaults standardUserDefaults] synchronize];
  
  [[NSNotificationCenter defaultCenter]
   postNotificationName:NYPLSettingsDidChangeNotification
   object:self];
}

- (void)setCurrentCardApplication:(NYPLCardApplicationModel *)currentCardApplication
{
  if (!currentCardApplication) {
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:currentCardApplicationSerializationKey];
    return;
  }
  
  NSData *cardAppData = [NSKeyedArchiver archivedDataWithRootObject:currentCardApplication];
  
  [[NSUserDefaults standardUserDefaults] setObject:cardAppData forKey:currentCardApplicationSerializationKey];
  [[NSUserDefaults standardUserDefaults] synchronize];
  
  [[NSNotificationCenter defaultCenter]
   postNotificationName:NYPLSettingsDidChangeNotification
   object:self];
}

- (void)setSettingsAccountsList:(NSArray *)accounts
{
  [[NSUserDefaults standardUserDefaults] setObject:accounts forKey:settingsLibraryAccountsKey];
  [[NSUserDefaults standardUserDefaults] synchronize];
}

- (NYPLSettingsRenderingEngine)renderingEngine
{
  return RenderingEngineFromString([[NSUserDefaults standardUserDefaults]
                                    stringForKey:renderingEngineKey]);
}

- (void)setRenderingEngine:(NYPLSettingsRenderingEngine const)renderingEngine
{
  if (renderingEngine == self.renderingEngine) return;
  
  [[NSUserDefaults standardUserDefaults] setObject:StringFromRenderingEngine(renderingEngine)
                                            forKey:renderingEngineKey];
  
  [[NSNotificationCenter defaultCenter]
   postNotificationName:NYPLSettingsDidChangeNotification
   object:self];
}

- (void)migrate
{
  NSInteger version = [[NSUserDefaults standardUserDefaults] integerForKey:versionKey];
  while (version < Version) {
    switch (version) {
      case 0: {
        NSArray *libraryAccounts = [[NSUserDefaults standardUserDefaults] arrayForKey:settingsLibraryAccountsKey];
        if (!libraryAccounts) {
          break;
        }
        
        // Load Accounts.json
        NSString *filePath = [[NSBundle mainBundle] pathForResource:@"Accounts" ofType:@"json"];
        if (filePath) {
          NSData *data = [NSData dataWithContentsOfFile:filePath];
          __autoreleasing NSError* error = nil;
          NSArray *accountList = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&error];
          NSMutableDictionary *idToUuidMap = [[NSMutableDictionary alloc] init];
          for (NSDictionary *account in accountList) {
            id numericVal = [account objectForKey:@"id_numeric"];
            id uuidVal = [account objectForKey:@"id_uuid"];
            if (numericVal && uuidVal) {
              [idToUuidMap setObject:uuidVal forKey:[numericVal stringValue]];
            }
          }
          
          // Migrate accounts
          NSMutableArray* newLibraryAccountsList = [NSMutableArray arrayWithCapacity:libraryAccounts.count];
          for (id account in libraryAccounts) {
            id uuidObj = [idToUuidMap objectForKey:[account stringValue]];
            if (uuidObj) {
              [newLibraryAccountsList addObject:uuidObj];
            }
          }
          [[NSUserDefaults standardUserDefaults] setObject:newLibraryAccountsList forKey:settingsLibraryAccountsKey];
        } else {
          [[NSUserDefaults standardUserDefaults] removeObjectForKey:settingsLibraryAccountsKey];
        }
        [[NSUserDefaults standardUserDefaults] synchronize];
      }
        break;
      default:
        break;
    }
    version += 1;
  }
  [[NSUserDefaults standardUserDefaults] setInteger:Version forKey:versionKey];
}

@end

File: Simplified/NYPLSettings.swift
import Foundation

extension Notification.Name {
  static let NYPLSettingsDidChange = Notification.Name("NYPLSettingsDidChange")
  static let NYPLCurrentAccountDidChange = Notification.Name("NYPLCurrentAccountDidChange")
  static let NYPLCatalogDidLoad = Notification.Name("NYPLCatalogDidLoad")
  static let NYPLSyncBegan = Notification.Name("NYPLSyncBegan")
  static let NYPLSyncEnded = Notification.Name("NYPLSyncEnded")
  static let NYPLUseBetaDidChange = Notification.Name("NYPLUseBetaDidChange")
}

@objc extension NSNotification {
  public static let NYPLSettingsDidChange = Notification.Name.NYPLSettingsDidChange
  public static let NYPLCurrentAccountDidChange = Notification.Name.NYPLCurrentAccountDidChange
  public static let NYPLCatalogDidLoad = Notification.Name.NYPLCatalogDidLoad
  public static let NYPLSyncBegan = Notification.Name.NYPLSyncBegan
  public static let NYPLSyncEnded = Notification.Name.NYPLSyncEnded
  public static let NYPLUseBetaDidChange = Notification.Name.NYPLUseBetaDidChange
}

let Version = 1

@objcMembers class NYPLSettings: NSObject {
  static let shared = NYPLSettings()
  
  static let NYPLAboutSimplyEURLString = "https://librarysimplified.org/simplye/"
  static let NYPLUserAgreementURLString = "https://www.librarysimplified.org/EULA/"
  
  static fileprivate let customMainFeedURLKey = "NYPLSettingsCustomMainFeedURL"
  static fileprivate let accountMainFeedURLKey = "NYPLSettingsAccountMainFeedURL"
  static fileprivate let userHasSeenWelcomeScreenKey = "NYPLUserHasSeenWelcomeScreenKey"
  static fileprivate let userPresentedAgeCheckKey = "NYPLUserPresentedAgeCheckKey"
  static fileprivate let userSeenFirstTimeSyncMessageKey = "userSeenFirstTimeSyncMessageKey"
  static fileprivate let useBetaLibrariesKey = "NYPLUseBetaLibrariesKey"
  static fileprivate let settingsLibraryAccountsKey = "NYPLSettingsLibraryAccountsKey"
  static fileprivate let versionKey = "NYPLSettingsVersionKey"
  
  @objc class func sharedSettings() -> NYPLSettings {
    return NYPLSettings.shared
  }

  // used to handle saml and oauth login. In case of the later, it needs to be configured as universal link
  @objc var authenticationUniversalLink: URL = URL(string: "https://librarysimplified.org/login")!

  // Set to nil (the default) if no custom feed should be used.
  var customMainFeedURL: URL? {
    get {
      return UserDefaults.standard.url(forKey: NYPLSettings.customMainFeedURLKey)
    }
    set(customUrl) {
      if (customUrl == self.customMainFeedURL) {
        return
      }
      UserDefaults.standard.set(customUrl, forKey: NYPLSettings.customMainFeedURLKey)
      UserDefaults.standard.synchronize()
      NotificationCenter.default.post(name: Notification.Name.NYPLSettingsDidChange, object: self)
    }
  }
  
  var accountMainFeedURL: URL? {
    get {
      return UserDefaults.standard.url(forKey: NYPLSettings.accountMainFeedURLKey)
    }
    set(mainFeedUrl) {
      if (mainFeedUrl == self.accountMainFeedURL) {
        return
      }
      UserDefaults.standard.set(mainFeedUrl, forKey: NYPLSettings.accountMainFeedURLKey)
      UserDefaults.standard.synchronize()
      NotificationCenter.default.post(name: Notification.Name.NYPLSettingsDidChange, object: self)
    }
  }
  
  var userHasSeenWelcomeScreen: Bool {
    get {
      return UserDefaults.standard.bool(forKey: NYPLSettings.userHasSeenWelcomeScreenKey)
    }
    set(b) {
      UserDefaults.standard.set(b, forKey: NYPLSettings.userHasSeenWelcomeScreenKey)
      UserDefaults.standard.synchronize()
    }
  }
  
  var userPresentedAgeCheck: Bool {
    get {
      return UserDefaults.standard.bool(forKey: NYPLSettings.userPresentedAgeCheckKey)
    }
    set(b) {
      UserDefaults.standard.set(b, forKey: NYPLSettings.userPresentedAgeCheckKey)
      UserDefaults.standard.synchronize()
    }
  }
  
  var userHasSeenFirstTimeSyncMessage: Bool {
    get {
      return UserDefaults.standard.bool(forKey: NYPLSettings.userSeenFirstTimeSyncMessageKey)
    }
    set(b) {
      UserDefaults.standard.set(b, forKey: NYPLSettings.userSeenFirstTimeSyncMessageKey)
      UserDefaults.standard.synchronize()
    }
  }
  
  var useBetaLibraries: Bool {
    get {
      return UserDefaults.standard.bool(forKey: NYPLSettings.useBetaLibrariesKey)
    }
    set(b) {
      UserDefaults.standard.set(b, forKey: NYPLSettings.useBetaLibrariesKey)
      UserDefaults.standard.synchronize()
      NotificationCenter.default.post(name: NSNotification.Name.NYPLUseBetaDidChange, object: self)
    }
  }
  
  var settingsAccountsList: [String] {
    get {
      if let libraryAccounts = UserDefaults.standard.array(forKey: NYPLSettings.settingsLibraryAccountsKey) {
        return libraryAccounts as! [String]
      }
      
      // Avoid crash in case currentLibrary isn't set yet
      var accountsList = [String]()
      if let currentLibrary = AccountsManager.shared.currentAccount?.uuid {
        accountsList.append(currentLibrary)
      }
      accountsList.append(AccountsManager.NYPLAccountUUIDs[2])
      self.settingsAccountsList = accountsList
      return accountsList
    }
    set(newAccountsList) {
      UserDefaults.standard.set(newAccountsList, forKey: NYPLSettings.settingsLibraryAccountsKey)
      UserDefaults.standard.synchronize()
    }
  }
  
  var appVersion: String? {
    get {
      return UserDefaults.standard.string(forKey: NYPLSettings.versionKey)
    }
    set(versionString) {
      UserDefaults.standard.set(versionString, forKey: NYPLSettings.versionKey)
      UserDefaults.standard.synchronize()
    }
  }
}
